Fix testQuotedQueryStringWithBoost test

mayya-sharipova opened the pull request:
Seems that asynchronous indexing from indexRandom sometimes indexes
the same document twice, which will mess up the expected score calculations.

For example, indexing:
```js
{ "index" : {"_id" : "1" } }
{"important" :"phrase match", "less_important": "nothing important"}
{ "index" : {"_id" : "2" } }
{"important" :"nothing important", "less_important" :"phrase match"}
```
Produces the expected scores: 13.8 for doc1, and 1.38 for doc2

indexing:
```js
{ "index" : {"_id" : "1" } }
{"important" :"phrase match", "less_important": "nothing important"}
{ "index" : {"_id" : "2" } }
{"important" :"nothing important", "less_important" :"phrase match"}
{ "index" : {"_id" : "3" } }
{"important" :"phrase match", "less_important": "nothing important"}
```
Produces scores: 9.4 for doc1, and 1.96 for doc2 which are found in the error logs.

I am not sure exactly how this happens, but fairly confident that
removing indexRandom should solve this test failure.

Closes #43144

mayya-sharipova committed:
Fix testQuotedQueryStringWithBoost test

Seems that asynchronous indexing from indexRandom sometimes indexes
the same document twice, which will mess up the expected score calculations.

For example, indexing:
{ "index" : {"_id" : "1" } }
{"important" :"phrase match", "less_important": "nothing important"}
{ "index" : {"_id" : "2" } }
{"important" :"nothing important", "less_important" :"phrase match"}
Produces the expected scores: 13.8 for doc1, and 1.38 for doc2

indexing:
{ "index" : {"_id" : "1" } }
{"important" :"phrase match", "less_important": "nothing important"}
{ "index" : {"_id" : "2" } }
{"important" :"nothing important", "less_important" :"phrase match"}
{ "index" : {"_id" : "3" } }
{"important" :"phrase match", "less_important": "nothing important"}
Produces scores: 9.4 for doc1, and 1.96 for doc2 which are found in the
error logs.

I am not sure exactly how this happens, but fairly confident that
removing indexRandom should solve this test failure.

Closes #43144

elasticmachine commented:
Pinging @elastic/es-search

cbuescher commented on server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java:
```diff
@@ -793,28 +793,20 @@ public void testFuzzyQueryString() {
-        assertHitCount(searchResponse, 2L);
-        assertFirstHit(searchResponse, hasId("1"));
-        assertSecondHit(searchResponse, hasId("2"));
-        assertThat((double)searchResponse.getHits().getAt(0).getScore(), closeTo(boost * searchResponse.getHits().getAt(1).getScore(), .1));
```
Nice delete, I also could not spot any difference here.

cbuescher reviewed (changes requested):
@mayya-sharipova great detective work, I just tried your example indexing doc1 twice and the numbers really do add up. Whats worrying me though is that I also cannot spot any way that indexRandom() might cause this, especially since the doc ids are provided and I don't see how that helper method would create a third document.
Another this that odd: when I run your re-creation and the search, I get:
```
"hits" : [
      {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : 9.400073,
        "_source" : {
          "important" : "phrase match",
          "less_important" : "nothing important"
        }
      },
      {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "3",
        "_score" : 9.400073,
        "_source" : {
          "important" : "phrase match",
          "less_important" : "nothing important"
        }
      },
      {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : 1.9616584,
        "_source" : {
          "important" : "nothing important",
          "less_important" : "phrase match"
        }
      }
    ]
```
and the test in L927 (7.x) compares `getHits().getAt(0).getScore()` and `getHits().getAt(1).getScore()`, but with two similar documents those would be the same? Actually I see there is a check earlier that the result set is of size 2, so that makes the "third" document scenario kind of weird.

I saw that "indexRandom()" in its full form (thats called eventually by the call in the test) has a `dummyDocuments` flag (`public void indexRandom(boolean forceRefresh, boolean dummyDocuments, boolean maybeFlush, List<IndexRequestBuilder> builders)` that does some weird tests around deleted documents. I'd suspect that to be the culprit. Maybe we can first try to call `indexRandom(boolean forceRefresh, boolean dummyDocuments, List<IndexRequestBuilder> builders)` with `forceRefresh` set to `true` but `dummyDocuments` to false?
My preference would be to instead of removing "indexRandom" (which should work as expected, otherwise we need to fix that part) to use it without dummy documents.

mayya-sharipova committed:
Add more logging to debug test case

mayya-sharipova commented:
@cbuescher  Thanks for looking into this, and sorry for a late reply.

>  to use it without dummy documents

I am not sure how dummy documents can influence scoring, as they are empty documents and don't contain any fields.. And the error seem to also happen when there are no dummy documents, for example in this case (notice we are indexing only 2 of our original documents):
```
1> [2019-02-16T06:42:08,639][INFO ][o.e.s.q.SearchQueryIT    ] [testQuotedQueryStringWithBoost] Index [2] docs async: [true] bulk: [false]
  1> [2019-02-16T06:42:08,642][INFO ][o.e.c.m.MetaDataMappingService] [node_s1] [test/RSRH3OjBRDyB-qsI-VuGyA] create_mapping [type1]
  1> [2019-02-16T06:42:08,666][INFO ][o.e.c.r.a.AllocationService] [node_s1] Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[test][0]] ...]).
  1> [2019-02-16T06:42:08,682][INFO ][o.e.s.q.SearchQueryIT    ] [testQuotedQueryStringWithBoost] [SearchQueryIT#testQuotedQueryStringWithBoost]: cleaning up after test
  1> [2019-02-16T06:42:08,692][INFO ][o.e.c.m.MetaDataDeleteIndexService] [node_s1] [test/RSRH3OjBRDyB-qsI-VuGyA] deleting index
  1> [2019-02-16T06:42:08,711][INFO ][o.e.c.m.MetaDataIndexTemplateService] [node_s1] removing template [random_index_template]
  1> [2019-02-16T06:42:08,717][INFO ][o.e.s.q.SearchQueryIT    ] [testQuotedQueryStringWithBoost] [SearchQueryIT#testQuotedQueryStringWithBoost]: cleaned up after test
  1> [2019-02-16T06:42:08,717][INFO ][o.e.s.q.SearchQueryIT    ] [testQuotedQueryStringWithBoost] after test
FAILURE 0.17s J7 | SearchQueryIT.testQuotedQueryStringWithBoost <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: 
   > Expected: a numeric value within <0.1> of <9.400073051452637>
   >      but: <19.61658477783203> differed by <10.116511726379395>
```

But just in case, I have modified the test case to use `indexRandom` without any dummy documents.

I wonder if errors happen because some documents are reindexed two times. I have added another log line to help us debugging.  WDYT, @cbuescher ?

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into fix-text-quoted-querystring-with-boost

cbuescher commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -1361,6 +1361,8 @@ public void indexRandom(boolean forceRefresh, boolean dummyDocuments, boolean ma
         for (Tuple<IndexRequestBuilder, Exception> tuple : errors) {
             if (ExceptionsHelper.unwrapCause(tuple.v2()) instanceof EsRejectedExecutionException) {
+                // temporarily logging errors to aid debugging https://github.com/elastic/elasticsearch/issues/43144
+                logger.info("Error indexing, reindexing doc");
```
Maybe it would it make sense to add more details (like the message) of the exception we unwrapped here to get even more info about whats happening.
I think it makes sense to have this logging in general (not only to investigate 43144) but I'm wondering if it would make sense to set this to a lower level (e.g. Debug) and enable it on the test we are investigating in 43144 using the new `@TestIssueLogging` annotation that is just introduced in https://github.com/elastic/elasticsearch/pull/44477). Might be worth waiting for that to be merged, otherwise you could also use the existing `@TestLogging` annotation to set the level to DEBUG for the test logging in `testQuotedQueryStringWithBoost()`.

cbuescher commented:
> I wonder if errors happen because some documents are reindexed two times. I have added another log line to help us debugging. WDYT

Great idea, I would rather get more information on this than work around the issue at this point. I left a comment around logging but I don't think the decision there needs another round of review, so LGTM.
I would keep #43144 open for a while though and revisit it in a few weeks just yet to see if we get more info if it fails again.

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into fix-text-quoted-querystring-with-boost

mayya-sharipova committed:
address feedback
